### DISCUSSION TITLE
[프론트엔드 배포 - p382] 삽질에 대한 공유 (feat. ElasticBeansTalk,Package.json, npm install)

### DISCUSSION THREAD

**Discussion by ghost** [fsoftwareengineer/todo-application#103] [질문과 답변] [answered]
### TL;DR
- 배포할때 node_modules디렉터리를 제공하지 않으면 [Amazone Linux 2 Node.js 플랫폼]은 npm install을 수행합니다. 
- 이때! t2.micro는 메모리가 부족해서 npm install하다가 죽습니다. 결국, 
1. node_modules를 포함시키거나, 
2. 메모리 덜먹는 패키지 매니저 툴을 쓰라고 [Amazone Linux 2 Node.js 플랫폼]에게 알려주거나( ex. yarn berry), 
3. 메모리 빵빵한 EC2를 써서 돈으로 해결하면 됩니다.(조심하세요! 다음달 요금 3000만원 나올 수 있어요)

### 그걸 어떻게 알았냐구요? 
- 1. 사망 로그를 뒤져보면 메모리가 없어서 죽은 기록이 있다. 
- 2. 공식 문서를 보고 확인하였습니다.(https://docs.aws.amazon.com/ko_kr/elasticbeanstalk/latest/dg/nodejs-platform-dependencies.html)
![Flynn 7](https://user-images.githubusercontent.com/12447259/160023724-e73a9628-725a-4497-8a97-43c108a381a6.png)
er-images.githubusercontent.com/12447259/160019847-72058b99-effc-4f36-aaf4-6f2b245d38e2.png)

### 작가님 이거 왜 안되요? 시키는 대로 했는데! 자꾸 Severe에요!! - 삽질 공유 ###

콘솔에는 이런 메세지가 뜨고
![Flynn 8](https://user-images.githubusercontent.com/12447259/160020715-e848cafb-63c2-44be-a2d5-081570933a22.png)
아마존 콘솔에는 
![Flynn 9](https://user-images.githubusercontent.com/12447259/160020853-bc6bb069-34e3-4c19-85cf-bbc962f992c0.png)
이런 모습이다가 Severe로 상태가 바뀝니다.

### 이유 없는 무덤은 없는 법 ###
- 로그-> 전체로그를 다운로드 후 이유를 찾아봅니다. eb-engine.log 파일 입니다.
![Flynn 5](https://user-images.githubusercontent.com/12447259/160018211-f8361f61-a099-4cbf-9ef4-6cec869531b6.png)

잘 수행 하다가 그냥 명령어 하나 띄웠는데 죽는다? 이건 OS 문제지!!! 커널 로그로 갑니다.
(그 명령어는 process이고, 우리의 경우 npm install 입니다.)

아마존은 친절하게도 사망하기 전에 페이지 사용 내역을 한방 띄워주고 프로세스를 죽이고 커널이 죽지 않게 합니다.(사실 다른 OS들도..)
![Flynn 6](https://user-images.githubusercontent.com/12447259/160024534-76f654f7-39e8-44a8-862d-58629ffa85ef.png)

_**아! 알았다. npm install 네가 문제 였구나!**_

### 해결법 ###
1. node_modules를 포함시키거나, 
2. 메모리 덜먹는 패키지 매니저 툴을 쓰라고 [Amazone Linux 2 Node.js 플랫폼] 알려주거나( ex-yarn berry), 
3. 메모리 빵빵한 EC2를 써서 돈으로 해결하면 됩니다.(조심하세요! 다음달 요금 3000만원 나올 수 있어요)

쉽고 간단한, 1번으로 해결합니다. 
- 2번의 해결법은 검색 하거나 스택오버 플로우 뒤진 결과는 .ebextion에서 옵션을 주거나, 아마존 리눅스 플랫폼 hook에서 지정하는 방법이 있다고 하는데.. 아직 아마존 LINUX에 대한 공부가 깊지 않아서 ㅠㅠ.  사실 2번 방법이 완벽한 해결법이라고 생각합니다. CD환경에서는 1번은 못 쓰는 방법이라고 생각해요. 검색 해보시면 옛날 기록들이 많아서 config 파일안에 file: 로 이그젝할 파일들 지정 해주는 스크립트(?)들이 있는데, 그건 아마존 리눅스 1버전입니다. 2버전을 찾아야하는데, 찾아도 잘 나오지 않고, 몇몇이 나와서 시도해보았지만 실패 했습니다. 결국 아마존 리눅스를 잘 알아야한다는 결론에 도달했습니다. 어쨋든 배포는 해야하니 1번으로 문제 해결 방향을 바꿨습니다.
- 3번의 해결법은 돈이 드니까 싫어요.

1번 해결법
배포할때 node_modules를 포함시키면 되겠죠?
프로젝트 루트에서 .gitignore와 같은레벨에서 
1. .ebignore 파일을 만들고
`.git`와 `.gitignore` 를 써줍니다. 
(사실 좀 더 세밀하게 이그노어 디렉토리를 지정해야 하지만.. 아마존 리눅스 플랫폼이 또 뭘 땡겨 쓸지 몰라서.. ㅠㅠ )

2. 배포하시면 됩니다.아마존 콘솔에서 프론트엔드 환경과 어플리케이션을 삭제 후 eb create -> eb deploy. 
![Flynn 10](https://user-images.githubusercontent.com/12447259/160023921-090626d1-386b-4650-9557-c3d5e38d05f3.png)
![Flynn 11](https://user-images.githubusercontent.com/12447259/160023932-38035248-3e60-4393-82ae-810295d31627.png)
3. 책에서 시키는대로 CORS문제 해결하러 가시면 됩니다.

### 미세먼지의 계절이니, 미세 먼지 Tip
package.json에 
	"engines": {
		"node": ">=15 <17"
	},
추가 하시면, [아마존 리눅스 2 node.js 플랫폼]은 해당 범위 내의 노드 js중 최신을 설치 합니다.
![Flynn 12](https://user-images.githubusercontent.com/12447259/160024004-2a3dc81f-0b40-4c34-a7b1-7ff32cc93fb5.png)

### 2번 해결법에 대한 의견을 듣고싶어요 ###
2번 해결법의 핵심은 아마존 리눅스 플랫폼이 node_modules가 없을때 수행하는 명령어를 npm install 이아니라,  아마존 리눅스에 설치된 yarn berry 버전으로 install 하고, run 하는건데. 이런 문제는 어떻게 풀죠? 무엇을 공부해야 할까요? 아시는 분 댓글 남겨주세요. 아마존에서 쓰는 yml과 config 파일들이 너무 어렵게 느껴집니당..

-끝 - 
ref) https://dev.to/mweibel/deploying-next-js-on-aws-elasticbeanstalk-ab4
ref) https://docs.aws.amazon.com/ko_kr/elasticbeanstalk/latest/dg/nodejs-platform-dependencies.html

**Comment by fsoftwareengineer** [comment, marked as answer]
쉽지 않은 디버깅 과정을 공유 해 주시고 좋은 문서로 남겨 주셔서 감사합니다.

**Comment by wch18735** [comment]
이렇게 진행했을 때, permission denied가 나타나는 문제는 없으셨나요?

```log
web: sh: /var/app/current/node_modules/.bin/react-scripts: Permission denied
```

**Reply by ghost** [reply]
너무 늦게 댓글을 봐서 이미 해결 하셨을 것 같지만, 다른분들을 위해 남겨요.
노드모듈을 통째로 업로드 하고 아마존 리눅스 위에서는 리액트가 노드 모듈을 참조 합니다. 
이때 디렉토리에 권한이 풀려있지 않다면, 퍼미션 디나인드가 뜰 수 있어요.
755 혹은 644 등 적절한 권한설정을 해주길 바랍니다. (참조할 검색어 : 리눅스 파일 권한 설정)
